stan/math/prim/fun/generalized_inverse.hpp

```diff
@@ -0,0 +1,68 @@
+#ifndef STAN_MATH_PRIM_FUN_GENERALIZED_INVERSE_HPP
+#define STAN_MATH_PRIM_FUN_GENERALIZED_INVERSE_HPP
+
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/prim/err.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/fun/chol2inv.hpp>
+#include <stan/math/prim/fun/add_diag.hpp>
+#include <stan/math/prim/fun/rep_vector.hpp>
+#include <stan/math/prim/fun/cholesky_decompose.hpp>
+#include <stan/math/prim/fun/transpose.hpp>
+#include <stan/math/prim/fun/tcrossprod.hpp>
+#include <stan/math/prim/fun/crossprod.hpp>
+#include <stan/math/prim/fun/multiply_lower_tri_self_transpose.hpp>
+#include <stan/math/prim/fun/mdivide_left_tri_low.hpp>
+
+namespace stan {
+namespace math {
+
+/**
+ * Returns the Moore-Penrose generalized inverse of the specified matrix.
+ *
+ * @tparam T type of elements in the matrix
+ * @tparam n number of rows, can be Eigen::Dynamic
+ * @tparam m number of columns, can be Eigen::Dynamic
+ *
+ * @param G specified matrix
+ * @return Generalized inverse of the matrix (an empty matrix if the specified
+ * matrix has size zero).
+ */
+template <typename EigMat,
+          require_eigen_vt<std::is_arithmetic, EigMat>* = nullptr>
+inline Eigen::Matrix<value_type_t<EigMat>, EigMat::RowsAtCompileTime,
+                     EigMat::ColsAtCompileTime>
+generalized_inverse(const EigMat& G) {
+  using value_t = value_type_t<EigMat>;
+  if (G.size() == 0) {
+    return {};
+  }
+
+  const auto n = G.rows();
+  const auto m = G.cols();
+
+  if (G.rows() == G.cols()) {
+    return G.inverse();
+  }
+
+  if (n < m) {
+    Eigen::Matrix<value_t, -1, -1> A = tcrossprod(G);
+    A.diagonal().array() += Eigen::Array<double, -1, 1>::Constant(n, 1e-10);
```

Comment:
If matrix contains only small values this can be a large relative change, which can affect the result a lot. Would it make sense to instead do something like? 
```
A.diagonal().array() *= 1 + 1e-10;
```

Reply:
I think it's even better if we let the user handle the constant. In the test I was able to get 1e-9 accuracy finding the appropriate constant. For matrices where the Cholesky decomposition is successful one does not need to add this jitter.